### PR TITLE
bpo-31926: fix missing *_METHODDEF statements by argument clinic

### PR DESCRIPTION
**Pull request by taleinat**
When a single .c file contains several functions and/or methods with the same name, a safety `_METHODDEF` `#define` block is generated only for one of them.

This fixes the bug by using the full name of the function to avoid duplicates rather than just the name.

https://bugs.python.org/issue31926

### CODE REVIEW

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by vstinner** [comment]
You have to run "make clinic": PR must included generated files.

Running "make clinic" changes Modules/clinic/zlibmodule.c.h to add:

```
diff --git a/Modules/clinic/zlibmodule.c.h b/Modules/clinic/zlibmodule.c.h
index 3edf7db7c2..33c767236b 100644
--- a/Modules/clinic/zlibmodule.c.h
+++ b/Modules/clinic/zlibmodule.c.h
@@ -467,4 +467,8 @@ exit:
 #ifndef ZLIB_COMPRESS_COPY_METHODDEF
     #define ZLIB_COMPRESS_COPY_METHODDEF
 #endif /* !defined(ZLIB_COMPRESS_COPY_METHODDEF) */
-/*[clinic end generated code: output=e0184313eb431e95 input=a9049054013a1b77]*/
+
+#ifndef ZLIB_DECOMPRESS_COPY_METHODDEF
+    #define ZLIB_DECOMPRESS_COPY_METHODDEF
+#endif /* !defined(ZLIB_DECOMPRESS_COPY_METHODDEF) */
+/*[clinic end generated code: output=6378d429f0819817 input=a9049054013a1b77]*/
```

So I think that you can remove the #ifdef in zlibmodule.c:

```
static PyMethodDef Decomp_methods[] =
{
    ZLIB_DECOMPRESS_DECOMPRESS_METHODDEF
    ZLIB_DECOMPRESS_FLUSH_METHODDEF
#ifdef HAVE_ZLIB_COPY
    ZLIB_DECOMPRESS_COPY_METHODDEF
#endif
    {NULL, NULL}
};
```

**Comment by serhiy-storchaka** [comment]
I think it is worth to add a news entry into the "Build" section.

**Comment by vstinner** [comment]
Oh, and you can do the same change for ZLIB_COMPRESS_COPY_METHODDEF.

**Comment by taleinat** [comment]
I've done all that @serhiy-storchaka and @haypo suggested.

Please review the `NEWS` entry I wrote.

Note: This will require manual backporting.

File: Misc/NEWS.d/next/Build/2017-11-02-23-10-08.bpo-31926.tbb5xH.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+rgument Clinic failing to generate some ``#ifndef ... #define`` blocks for
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Build/2017-11-02-23-10-08.bpo-31926.tbb5xH.rst]: :scissors:

Comment (taleinat) [reply Misc/NEWS.d/next/Build/2017-11-02-23-10-08.bpo-31926.tbb5xH.rst]: Done.

Can this go in?

Comment (serhiy-storchaka) [reply Misc/NEWS.d/next/Build/2017-11-02-23-10-08.bpo-31926.tbb5xH.rst]: Sorry for been too terse and ambiguous. I meant that the NEWS entry were cut (missed starting "A" and likely "Fixed"), not that it should be shortened. But in any case the new wording looks better to me!

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by taleinat** [comment]
Should I make the 3.6 backport?

**Comment by serhiy-storchaka** [comment]
Yes, please do this. We will not convert new files to Argument Clinic in 3.6, and the existing files contain workarounds, but it is better to fix the bug.

**Pull request merged**